include/range/v3/view/for_each.hpp

```diff
@@ -58,35 +46,11 @@ namespace ranges
                 )
             public:
                 template<typename Rng, typename F>
-                using Concept = meta::and_<
-                    Range<Rng>,
-                    IndirectInvocable<F, iterator_t<Rng>>,
-                    Range<concepts::Invocable::result_t<F&, range_common_reference_t<Rng>>>>;
-
-                template<typename Rng, typename F,
-                    CONCEPT_REQUIRES_(Concept<Rng, F>())>
-                for_each_view<all_t<Rng>, F> operator()(Rng && rng, F f) const
-                {
-                    return {all(std::forward<Rng>(rng)), std::move(f)};
-                }
-
-            #ifndef RANGES_DOXYGEN_INVOKED
-                // For better error reporting
-                template<typename Rng, typename F,
-                    CONCEPT_REQUIRES_(!Concept<Rng, F>())>
-                void operator()(Rng &&, F) const
-                {
-                    CONCEPT_ASSERT_MSG(Range<Rng>(),
-                        "Rng is not a model of the Range concept.");
-                    CONCEPT_ASSERT_MSG(IndirectInvocable<F, iterator_t<Rng>>(),
-                        "The function F is not callable with arguments of the type of the range's "
-                        "common reference type.");
-                    CONCEPT_ASSERT_MSG(Range<concepts::Invocable::result_t<F&,
-                        range_common_reference_t<Rng>>>(),
-                        "To use view::for_each, the function F must return a model of the Range "
-                        "concept.");
-                }
-            #endif
+                auto operator()(Rng && rng, F f) const
+                RANGES_DECLTYPE_AUTO_RETURN
+                (
+                    join(transform(std::forward<Rng>(rng), std::move(f)))
+                )
```

Comment:
No, Casey, that won't work. `join` and `transform` accept everything and have their own "bad" overloads of `operator()` that will diagnose failures. FloopCZ has done exactly the right thing here by not even constraining the parameters to `for_each_fn::operator()`, and just passing them through.